From: Suren Baghdasaryan <surenb@google.com>
Subject: kernel/fork: assert no VMA readers during its destruction
Date: Mon, 27 Feb 2023 09:36:19 -0800

Assert there are no holders of VMA lock for reading when it is about to be
destroyed.

Link: https://lkml.kernel.org/r/20230227173632.3292573-21-surenb@google.com
Signed-off-by: Suren Baghdasaryan <surenb@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---


--- a/kernel/fork.c~kernel-fork-assert-no-vma-readers-during-its-destruction
+++ a/kernel/fork.c
@@ -491,6 +491,9 @@ static void vm_area_free_rcu_cb(struct r
 {
 	struct vm_area_struct *vma = container_of(head, struct vm_area_struct,
 						  vm_rcu);
+
+	/* The vma should not be locked while being destroyed. */
+	VM_BUG_ON_VMA(rwsem_is_locked(&vma->lock), vma);
 	__vm_area_free(vma);
 }
 #endif
_
